service/history/workflow/workflow_task_state_machine.go

```diff
@@ -1087,11 +1080,64 @@ func (m *workflowTaskStateMachine) afterAddWorkflowTaskCompletedEvent(
 	attrs := event.GetWorkflowTaskCompletedEventAttributes()
 	m.ms.executionInfo.LastCompletedWorkflowTaskStartedEventId = attrs.GetStartedEventId()
 	m.ms.executionInfo.MostRecentWorkerVersionStamp = attrs.GetWorkerVersion()
-	if err := m.ms.CompleteDeploymentTransition(attrs.GetVersioningBehavior()); err != nil {
-		return err
+
+	wftDeployment := attrs.GetDeployment()
+	wftBehavior := attrs.GetVersioningBehavior()
+	versioningInfo := m.ms.GetExecutionInfo().GetVersioningInfo()
+
+	var completedTransition bool
+	if versioningInfo.DeploymentTransition != nil {
+		// It's possible that the completed WFT is not yet from the current transition because when
+		// the transition started, the current wft was already started. In this case, we allow the
+		// started wft to run and when completed, we create another wft immediately.
+		if versioningInfo.DeploymentTransition.GetDeployment().Equal(wftDeployment) {
+			versioningInfo.DeploymentTransition = nil
+			completedTransition = true
+		}
+	}
```

Comment:
`Equal` supports `nil` on both sides:
```suggestion
	// It's possible that the completed WFT is not yet from the current transition because when
	// the transition started, the current wft was already started. In this case, we allow the
	// started wft to run and when completed, we create another wft immediately.
	if versioningInfo.GetDeploymentTransition().GetDeployment().Equal(wftDeployment) {
		versioningInfo.DeploymentTransition = nil
		completedTransition = true
	}
```

Reply:
We don't want to set completedTransition to true if there was no transition and wftDeployment is nil.